README.md
# Dia

<a href="https://pub.dartlang.org/packages/dia">  
    <img src="https://img.shields.io/pub/v/dia.svg"  
      alt="Pub Package" />  
</a>
<a href="https://github.com/unger1984/dia">  
    <img alt="GitHub Repo stars" src="https://img.shields.io/github/stars/unger1984/dia">
</a>

A simple dart http server like KoaJS.

This package allows you to create a http / http server in a couple of lines. Dia creates a context from a bunch of request and response and passes it through the middleware.

The main idea of the project is minimalism. The package contains only basic functionality, everything else is implemented in separate packages. This allows you to keep the project code clean and connect only those dependencies that are really needed in it.

## Install:

Add to pubspec.yaml in dependencies section this:

```yaml
    dia: ^0.1.3
```

Then run `pub get`

## Usage:

A simple usage example:

```dart
import 'package:dia/dia.dart';

main() {
  final app = App();
  
  app.use((ctx,next) async {
    ctx.body = 'Hello world!';
  });

  app
      .listen('localhost', 8080)
      .then((info) => print('Server started on http://localhost:8080'));
}
```

Context contain getters and setters for `HttpRequest` fields: response,  response.headers, response.headers.contentType, response.statusCode, that allow use it easy.
Context contain method `throwError` that allow easy return HTTP errors by statusCode.

Example `throwError`:

```dart
    app.use((ctx,next) async {
      ctx.throwError(401);
    });
```

You can use custom `Context` with additional fields and methods. In this case, you need to inherit from the base Context and pass the instance of context creation function to the App constructor.

```dart
/// Create custom context class
class CustomContext extends Context{
  String? additionalField;
  CustomContext(HttpRequest request) : super(request);
}

void main() {
  /// Create Dia instance 
  final app = App((request) => CustomContext(request));

  /// Add additionalField value
  app.use((ctx, next) async {
    ctx.additionalField = 'additional value';
    await next();
  });

  /// final middleware to response
  app.use((ctx, next) async {
    ctx.contentType = ContentType.text;
    ctx.body = ctx.additionalField;
  });

  /// Start server listen on localhost:8080
  app
      .listen('localhost', 8080)
      .then((info) => print('Server started on http://localhost:8080'));
}
```

Your can add handler to all Http errors:

```dart
app.use((ctx,next) async {
  await next();
  if(ctx.statusCode!=200){
    //....
  }
});
```

You can start your server with SSL:

```dart
const serverKey = 'cert/key.pem';
const certificateChain = 'cert/chain.pem';

final serverContext = SecurityContext();
serverContext
    .useCertificateChainBytes(await File(certificateChain).readAsBytes());
serverContext.usePrivateKey(serverKey, password: 'password');

app.listen('localhost', 8444, securityContext: serverContext);
```

For more details, please, see example and test folder.

## Use with:

* [dia_router](https://github.com/unger1984/dia_router) - Package to route request as koa-router.
* [dia_cors](https://github.com/unger1984/dia_cors) - Package for CORS middleware.
* [dia_body](https://github.com/unger1984/dia_body) - Package with the middleware for parse request body.
* [dia_static](https://github.com/unger1984/dia_static) - Package to serving static files.

## Features and bugs:

I will be glad for any help and feedback!
Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/unger1984/dia/issues


CHANGELOG.md
# 0.1.3

- Fix error handling

# 0.1.2

- Fix error handling

## 0.1.1

- Change Error field in HttpError class to Exception.
- Add HttpError field in Context to handle http error before response
- Add handling error example to readme

## 0.1.0

- Remove dart:mirror, this now allows you to compile Dia in AOT mode.
- Add more information to readme

## 0.0.8

- Add `listenOn` method to listen http/https requests an existing ServerSocket
- Add more information to Readme
- Add more dartdoc comments in code

## 0.0.7

- Add default 404 answer
- Fix types

## 0.0.6

- Add link to dia_static
- Allow `Context.body` write `Stream`, `List` and `String`

## 0.0.5

- Change LICENSE

## 0.0.4

- Add `void close()` method, that close connection
- Add simple tests

## 0.0.3

- Fix links and descriptions

## 0.0.2

- Make `createInstance` public static in `Context`
- Add method `set(String key, String value)` to `Context` as symlink to `headers.set`
- Add the link to [dia_router](https://github.com/unger1984/dia_router)

## 0.0.1

- Initial version


test/custom.dart
import 'dart:io';

import 'package:http/http.dart' as http;
import 'package:test/test.dart';

// ignore: avoid_relative_lib_imports
import '../lib/dia.dart';

class CustomContext extends Context {
  String? field;
  CustomContext(HttpRequest request) : super(request);
}

void main() {
  App<CustomContext>? dia;

  setUp(() {
    dia = App((request) => CustomContext(request));
    dia?.listen('localhost', 8080);
  });

  tearDown(() async {
    dia?.close();
  });

  test('Dia serve custom context', () async {
    dia?.use((ctx, next) async {
      ctx.field = 'custom';
      await next();
    });

    dia?.use((ctx, next) async {
      ctx.body = ctx.field;
    });

    final response = await http.get(Uri.parse('http://localhost:8080'));
    expect(response.statusCode, equals(200));
    expect(response.body, equals('custom'));
  });
}


lib/src/app.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io' as io;

import 'context.dart';
import 'http_error.dart';

/// Middleware, it is an asynchronous function
/// that accepts a [Context] and a future for the next function.
typedef Middleware<T extends Context> = Future<void> Function(
  T ctx,
  FutureOr<void> Function() next,
);

/// Dia application class
/// The web server listens to the http / https port and applies
/// the middleware to the received requests
///
/// [Middleware] in App must be in the same [Context] as App
class App<T extends Context> {
  late io.HttpServer _server;
  final List<Middleware<T>> _middlewares = [];
  StreamSubscription? _listener;
  late Function(io.HttpRequest request) _createContext;

  App([T Function(io.HttpRequest request)? createContext]) {
    _createContext =
        createContext ?? (io.HttpRequest request) => Context(request);
  }

  /// Add [Middleware] to App
  /// [Middleware] must be in the same [Context] as App
  void use(Middleware<T> middleware) {
    _middlewares.add(middleware);
  }

  /// Call this method to listen http/https requests on the
  /// specified [address] and [port]
  ///
  /// The [address] can either be a [String] or an
  /// [InternetAddress]. If [address] is a [String], [bind] will
  /// perform a [InternetAddress.lookup] and use the first value in the
  /// list. To listen on the loopback adapter, which will allow only
  /// incoming connections from the local host, use the value
  /// [InternetAddress.loopbackIPv4] or
  /// [InternetAddress.loopbackIPv6]. To allow for incoming
  /// connection from the network use either one of the values
  /// [InternetAddress.anyIPv4] or [InternetAddress.anyIPv6] to
  /// bind to all interfaces or the IP address of a specific interface.
  ///
  /// If an IP version 6 (IPv6) address is used, both IP version 6
  /// (IPv6) and version 4 (IPv4) connections will be accepted. To
  /// restrict this to version 6 (IPv6) only, use [v6Only] to set
  /// version 6 only. However, if the address is
  /// [InternetAddress.loopbackIPv6], only IP version 6 (IPv6) connections
  /// will be accepted.
  ///
  /// If [port] has the value 0 an ephemeral port will be chosen by
  /// the system. The actual port used can be retrieved using the
  /// [port] getter.
  ///
  /// The optional argument [backlog] can be used to specify the listen
  /// backlog for the underlying OS listen setup. If [backlog] has the
  /// value of 0 (the default) a reasonable value will be chosen by
  /// the system.
  ///
  /// The optional argument [shared] specifies whether additional HttpServer
  /// objects can bind to the same combination of `address`, `port` and `v6Only`.
  /// If `shared` is `true` and more `HttpServer`s from this isolate or other
  /// isolates are bound to the port, then the incoming connections will be
  /// distributed among all the bound `HttpServer`s. Connections can be
  /// distributed over multiple isolates this way.
  Future<void> listen(
    address,
    int port, {
    io.SecurityContext? securityContext,
    int backlog = 0,
    bool v6Only = false,
    bool shared = false,
  }) async {
    _server = await (securityContext == null
        ? io.HttpServer.bind(
            address,
            port,
            backlog: backlog,
            v6Only: v6Only,
            shared: shared,
          )
        : io.HttpServer.bindSecure(
            address,
            port,
            securityContext,
            backlog: backlog,
            shared: shared,
          ));
    await _listen();
  }

  /// Call this method to listen http/https requests an
  /// existing [ServerSocket]. When the [HttpServer] is closed,
  /// the [HttpServer] will just detach itself, closing current
  /// connections but not closing [serverSocket].
  Future<void> listenOn(io.ServerSocket serverSocket) async {
    _server = io.HttpServer.listenOn(serverSocket);
    await _listen();
  }

  /// Close HTTP/HTTPS listener
  void close() {
    _listener?.cancel();
  }

  /// Private method for listen server stream
  Future<void> _listen() async {
    _listener = _server.listen((request) async {
      final ctx = _createContext(request);
      if (_middlewares.isNotEmpty) {
        final fn = _compose(_middlewares);
        await fn(ctx, null);
      }
      Stream<List<int>>? stream;
      if (ctx.body is String) {
        stream = Stream.fromIterable([utf8.encode(ctx.body)]);
      } else if (ctx.body is List) {
        stream = Stream.fromIterable([ctx.body]);
      } else if (ctx.body is Stream) {
        stream = ctx.body;
      }
      if (stream != null) {
        await request.response.addStream(stream);
      } else {
        request.response.headers.contentType = io.ContentType.html;
        request.response.statusCode = 404;
        request.response.write(HttpError(404).defaultBody);
      }
      await request.response.close();
    });
  }

  /// Private method for generate HTTP error response
  void _responseHttpError(T ctx, HttpError error) {
    ctx.error = error;
    ctx.statusCode = error.status;
    ctx.contentType = io.ContentType.html;
    ctx.body = error.defaultBody;
  }

  /// Private method for compose middlewares to one function
  Function _compose(List<Middleware<T>> middlewares) {
    return (T ctx, Middleware<T>? next) {
      var lastCalledIndex = -1;

      FutureOr dispatch(int currentCallIndex) async {
        if (currentCallIndex <= lastCalledIndex) {
          throw Exception('next() called multiple times');
        }
        lastCalledIndex = currentCallIndex;
        var fn = middlewares.length > currentCallIndex
            ? middlewares[currentCallIndex]
            : null;
        if (currentCallIndex == middlewares.length) {
          fn = next;
        }
        if (fn == null) return () => null;

        return fn(ctx, () => dispatch(currentCallIndex + 1))
            .catchError((error, stackTrace) {
          if (error is HttpError) {
            _responseHttpError(ctx, error);
          } else {
            final err = HttpError(
              500,
              stackTrace: stackTrace,
              exception: Exception(error),
            );
            _responseHttpError(ctx, err);
          }
        });
      }

      return dispatch(0);
    };
  }
}


lib/src/context.dart
import 'dart:io';

import 'http_error.dart';

/// Default Dia Context
/// This object extended base [HttpRequest] and
/// [HttpResponse] object for easy access to they fields
///
/// [request] - base [HttpRequest]
/// [response] - base [HttpResponse]
/// [headers] - base response [HttpHeaders]
/// [statusCode] - response status code
/// [contentType] - response [ContentType] field of headers
class Context {
  final HttpRequest _request;
  late dynamic body;

  /// Error if ctxThrow
  HttpError? error;

  /// Create base [Context] object
  /// takes [HttpRequest]
  Context(this._request) {
    // set default response content-type
    contentType ??= ContentType.html;
    // empty body
    body = null;
  }

  /// base [HttpRequest]
  HttpRequest get request => _request;

  /// base [HttpResponse]
  HttpResponse get response => _request.response;

  /// base response [HttpHeaders]
  HttpHeaders get headers => _request.response.headers;

  /// Response status code
  int get statusCode => response.statusCode;
  set statusCode(int code) => response.statusCode = code;

  /// Response Content-type field of headers
  ContentType? get contentType => headers.contentType;
  set contentType(ContentType? type) => headers.contentType = type;

  /// Set Response [HttpHeaders] field
  /// [key] - key of header
  /// [value] - value for key of header
  void set(String key, String value) {
    headers.set(key, value);
  }

  /// Throw HttpError with [status] code, [message] and [stackTrace] and [error]
  void throwError(int status,
      {String? message, StackTrace? stackTrace, Exception? exception}) {
    error = HttpError(status,
        message: message, stackTrace: stackTrace, exception: exception);
    throw error!;
  }
}


lib/src/http_error.dart
import 'dart:core';

/// HttpError class
/// Contain information about http error
/// [status] - response code
/// [message] - text message of code
/// [exception] - Exception objects thrown in the case of a program failure
/// [stackTrace] - [StackTrace] by all stack trace objects
class HttpError extends Error {
  final int _status;
  late String _message;
  late Exception? _exception;
  late final StackTrace? _stackTrace;

  /// Create [HttpError] object
  /// [status] - response code
  ///
  /// Optional parameters
  /// [message] - text message of code
  /// [exception] - Exception objects thrown in the case of a program failure
  /// [stackTrace] - [StackTrace] by all stack trace objects
  HttpError(
    this._status, {
    String? message,
    StackTrace? stackTrace,
    Exception? exception,
  }) : assert(_status >= 400 && _status <= 600,
            'The status should be an error code: 400-600') {
    _message = message ?? _codes[_status] ?? 'Unknown error';
    _stackTrace = stackTrace;
    _exception = exception;
  }

  /// text message of code
  String get message => _message;

  /// response code
  int get status => _status;

  /// Error objects thrown in the case of a program failure
  Exception? get exception => _exception;

  /// [StackTrace] by all stack trace objects
  @override
  StackTrace? get stackTrace => _stackTrace;

  /// Generate default HTML for this HTTP error
  /// without fail contains information about [status] and [message]
  /// additional, can contains information about [exception] and [stackTrace]
  String get defaultBody {
    var res = '''<html lang="en">
    <head>
      <title>$_status $_message</title>
    </head>
    <body>
      <h1>$_status $_message</h1>''';
    if (_exception != null) {
      res += '<h2>$_exception</h2>';
    }
    if (_stackTrace != null) {
      res += '''<h3>StackTrace</h3>
      <p>
      ''';
      final stack = _stackTrace.toString().split('\n');
      for (var index = 0; index < stack.length; index++) {
        final str = stack[index];
        res += '$str<br/>';
      }

      res += '</p>';
    }

    res += '''
    </body>
    </html>''';

    return res;
  }
}

/// http errors statuses and messages
const _codes = <int, String>{
  100: 'Continue',
  101: 'Switching Protocols',
  102: 'Processing',
  103: 'Early Hints',
  200: 'OK',
  201: 'Created',
  202: 'Accepted',
  203: 'Non-Authoritative Information',
  204: 'No Content',
  205: 'Reset Content',
  206: 'Partial Content',
  207: 'Multi-Status',
  208: 'Already Reported',
  226: 'IM Used',
  300: 'Multiple Choices',
  301: 'Moved Permanently',
  302: 'Found',
  303: 'See Other',
  304: 'Not Modified',
  305: 'Use Proxy',
  307: 'Temporary Redirect',
  308: 'Permanent Redirect',
  400: 'Bad Request',
  401: 'Unauthorized',
  402: 'Payment Required',
  403: 'Forbidden',
  404: 'Not Found',
  405: 'Method Not Allowed',
  406: 'Not Acceptable',
  407: 'Proxy Authentication Required',
  408: 'Request Timeout',
  409: 'Conflict',
  410: 'Gone',
  411: 'Length Required',
  412: 'Precondition Failed',
  413: 'Payload Too Large',
  414: 'URI Too Long',
  415: 'Unsupported Media Type',
  416: 'Range Not Satisfiable',
  417: 'Expectation Failed',
  418: "I'm a Teapot",
  421: 'Misdirected Request',
  422: 'Unprocessable Entity',
  423: 'Locked',
  424: 'Failed Dependency',
  425: 'Too Early',
  426: 'Upgrade Required',
  428: 'Precondition Required',
  429: 'Too Many Requests',
  431: 'Request Header Fields Too Large',
  451: 'Unavailable For Legal Reasons',
  500: 'Internal Server Error',
  501: 'Not Implemented',
  502: 'Bad Gateway',
  503: 'Service Unavailable',
  504: 'Gateway Timeout',
  505: 'HTTP Version Not Supported',
  506: 'Variant Also Negotiates',
  507: 'Insufficient Storage',
  508: 'Loop Detected',
  509: 'Bandwidth Limit Exceeded',
  510: 'Not Extended',
  511: 'Network Authentication Required',
};


example/dia_example.dart
import 'dart:io';

import 'package:dia/dia.dart';

/// Custom Context example
class MyContext extends Context {
  String? additionalField;

  MyContext(HttpRequest request) : super(request);
}

/// Middleware example
Middleware logger() => (ctx, next) async {
      var start = DateTime.now();
      await next();
      final time = DateTime.now().difference(start).inMilliseconds;
      print('${ctx.request.method} ${ctx.request.uri.path} ${time}ms');
    };

void main() {
  final app = App((request) => MyContext(request));

  /// Add logger middleware
  app.use(logger());

  /// Add additionalField value
  app.use((ctx, next) async {
    ctx.additionalField = 'additional value';
    await next();
  });

  /// Uncomment this to see throw example
  // app.use((ctx, next)async{
  //   ctx.throwError(401);
  // });

  /// final middleware to response
  app.use((ctx, next) async {
    ctx.contentType = ContentType.text;
    ctx.body = ctx.additionalField;
  });

  /// Start server listen on localhost:8080
  app
      .listen('localhost', 8080)
      .then((info) => print('Server started on http://localhost:8080'));
}
